### ISSUE TITLE
I think that headerValueByType does not work always with custom headers

### ISSUE THREAD

**Issue by xelax** [akka/akka#20203]
I tried headerValueByType with my own custom header but it does not seem to work. I put your own test in my own project and in that context it fails as well. Here is my test case, which fails with:

Run starting. Expected test count is: 1
ApiSpecs:
ApiSpecs
- should be able to extract in routing DSL via headerValueByType *** FAILED ***
  Request was rejected with rejection MissingHeaderRejection(ApiTokenHeader) (DynamicVariable.scala:58)
Run completed in 697 milliseconds.
Total number of tests run: 1
Suites: completed 1, aborted 0
Tests: succeeded 0, failed 1, canceled 0, ignored 0, pending 0
*** 1 TEST FAILED ***

```
import org.scalatest._
import akka.http.scaladsl.testkit.ScalatestRouteTest

import akka.http.scaladsl.model.headers.RawHeader
import akka.http.scaladsl.model.headers._
import akka.http.scaladsl.model._
import akka.http.scaladsl.server.MissingHeaderRejection
import scala.util.Try
import akka.http.scaladsl.server.Directives

object ApiToken {
  final class ApiTokenHeader(token: String) extends ModeledCustomHeader[ApiTokenHeader] {
    def renderInRequests = false
    def renderInResponses = false
    override val companion = ApiTokenHeader
    override def value: String = token
  }
  object ApiTokenHeader extends ModeledCustomHeaderCompanion[ApiTokenHeader] {
    def renderInRequests = false
    def renderInResponses = false
    override val name = "apiKey"
    override def parse(value: String) = Try(new ApiTokenHeader(value))
  }
}

class ApiSpecs extends FlatSpec with Matchers with Directives with ScalatestRouteTest {
  import ApiToken._

  "ApiSpecs" should "be able to extract in routing DSL via headerValueByType" in {
    val routes = headerValueByType[ApiTokenHeader]() { token ⇒
      val ApiTokenHeader(t) = token
      complete(s"extracted> $t")
    }

    Get().withHeaders(RawHeader("apiKey", "TheKey")) ~> routes ~> check {
      status should ===(StatusCodes.OK)
      responseAs[String] should ===("extracted> apiKey: TheKey")
    }

    Get().withHeaders(RawHeader("somethingElse", "TheKey")) ~> routes ~> check {
      rejection should ===(MissingHeaderRejection("ApiTokenHeader"))
    }

    Get().withHeaders(ApiTokenHeader("TheKey")) ~> routes ~> check {
      status should ===(StatusCodes.OK)
      responseAs[String] should ===("extracted> apiKey: TheKey")
    }
  }
}

```

**Comment by xelax** [comment]
tested with akka 2.3.15 and akka http stream 2.0.4

**Comment by ktoso** [comment]
This was fixed in 2.4.3, please upgrade.

**Comment by ktoso** [comment]
2.4.3 was released, contains fix: http://akka.io/news/2016/04/01/akka-2.4.3-released.html
Please upgrade, we won't be porting all fixes back to 2.0.x, only most critical ones.
